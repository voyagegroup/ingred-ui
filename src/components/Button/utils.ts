import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
