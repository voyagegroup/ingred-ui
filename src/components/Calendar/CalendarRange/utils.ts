import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
