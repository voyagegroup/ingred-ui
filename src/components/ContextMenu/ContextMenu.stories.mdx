import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import ContextMenu from "./ContextMenu";

<Meta
  title="Components/Navigation/ContextMenu"
  component={ContextMenu}
  argTypes={{ onClick: { action: "clicked" } }}
/>

# ContextMenu

<ArgsTable of={ContextMenu} />

## Samples

<Canvas>
  <Story
    name="Example"
    args={{
      contents: [
        {
          text: "Edit",
          onClick: () => {},
          type: "default",
        },
        {
          text: "Save",
          onClick: () => {},
          type: "default",
        },
      ],
    }}
  >
    {(args) => <ContextMenu {...args} />}
  </Story>
</Canvas>
