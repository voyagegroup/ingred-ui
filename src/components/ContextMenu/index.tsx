export { default } from "./ContextMenu";
export type { ContextMenuProps } from "./ContextMenu";
