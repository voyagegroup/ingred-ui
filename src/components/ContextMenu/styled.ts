import styled from "styled-components";
import OriginalActionButton from "../ActionButton";

export const ActionButton = styled(OriginalActionButton)`
  width: 34px;
  height: 34px;
`;
