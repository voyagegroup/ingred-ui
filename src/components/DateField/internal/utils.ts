export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
