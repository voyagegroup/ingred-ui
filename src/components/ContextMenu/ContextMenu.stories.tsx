import React from "react";
import { StoryObj } from "@storybook/react";
import { Title, ArgsTable, Stories } from "@storybook/addon-docs";
import { action } from "@storybook/addon-actions";
import ContextMenu, { ContextMenuProps } from "./ContextMenu";

export default {
  title: "Components/Navigation/ContextMenu",
  component: ContextMenu,
  parameters: {
    docs: {
      source: { language: "tsx" },
      page: () => (
        <>
          <Title />
          <ArgsTable of={ContextMenu} />
          <Stories includePrimary title="Stories" />
        </>
      ),
    },
  },
};

export const Basic: StoryObj<ContextMenuProps> = {
  args: {
    contents: [
      {
        text: "Edit",
        onClick: action('clicked "Edit"'),
      },
      {
        text: "Save",
        onClick: action('clicked "Save"'),
      },
    ],
  },
};

export const Disabled: StoryObj<ContextMenuProps> = {
  args: {
    contents: [
      {
        text: "Edit",
        onClick: action('clicked "Edit"'),
      },
      {
        text: "Save",
        onClick: () => {},
        type: "disabled",
      },
    ],
  },
};

export const Warning: StoryObj<ContextMenuProps> = {
  args: {
    contents: [
      {
        text: "Edit",
        onClick: action('clicked "Edit"'),
      },
      {
        text: "Delete",
        onClick: action('clicked "Delete"'),
        type: "warning",
      },
    ],
  },
};
