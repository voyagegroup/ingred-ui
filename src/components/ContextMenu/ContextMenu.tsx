import * as PopperJS from "@popperjs/core";
import * as React from "react";
import { useMergeRefs } from "../../hooks/useMergeRefs";
import { createChainedFunction } from "../../utils/createChainedFunction";
import { ActionButtonProps } from "../ActionButton";
import Menu, { MenuProps } from "../Menu";
import { ContentProp } from "../MenuList/MenuList";
import * as Styled from "./styled";

export type ContextMenuProps = {
  /**
   * `type ContentProp = React.ComponentPropsWithoutRef<"div"> & {
   *   text: string;
   *   onClick: () => void;
   *   divideTop?: boolean;
   *   disabled?: boolean;
   * }`
   */
  contents: ContentProp[];
  /**
   * Define priority of position. Please check [this](https://popper.js.org/docs/v2/modifiers/flip/#fallbackplacements).
   */
  positionPriority?: PopperJS.Placement[];
  menuMaxHeight?: MenuProps["maxHeight"];
  /**
   * props of [Menu](/?path=/docs/components-navigation-menu)
   */
  menuProps?: Partial<MenuProps>;
  /**
   * props of [ActionButton](/?path=/docs/components-inputs-actionbutton)
   */
  actionButtonProps?: Partial<ActionButtonProps>;
};

const ContextMenu = React.forwardRef<HTMLButtonElement, ContextMenuProps>(
  (
    {
      contents,
      positionPriority = ["bottom-start", "bottom-end", "top-start", "top-end"],
      menuMaxHeight = "none",
      menuProps,
      actionButtonProps,
    },
    ref,
  ) => {
    const [iconWrapperElement, setIconWrapperElement] =
      React.useState<HTMLButtonElement | null>(null);
    const [isOpen, setIsOpen] = React.useState<boolean>(false);

    const handleToggleOpen = (open: boolean) => () => {
      setIsOpen(open);
    };

    const refs = useMergeRefs<HTMLButtonElement>(setIconWrapperElement, ref);

    return (
      <>
        <Styled.ActionButton
          ref={refs}
          data-testid="icon-wrapper"
          icon="more_vert"
          onClick={handleToggleOpen(!isOpen)}
          {...actionButtonProps}
        />
        <Menu
          isOpen={isOpen}
          baseElement={iconWrapperElement}
          contents={contents}
          positionPriority={positionPriority}
          maxHeight={menuMaxHeight}
          {...menuProps}
          onClose={createChainedFunction(
            handleToggleOpen(false),
            menuProps?.onClose,
          )}
        />
      </>
    );
  },
);

export default ContextMenu;
