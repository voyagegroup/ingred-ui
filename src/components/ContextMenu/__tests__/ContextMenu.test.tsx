import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup, fireEvent, act } from "@testing-library/react";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";
import ContextMenu from "../ContextMenu";
import { ContentProp } from "../../MenuList/MenuList";

jest.mock("react-dom", () => {
  const original = jest.requireActual("react-dom");
  return {
    ...original,
    createPortal: (node: any) => node,
  };
});

const contents: ContentProp[] = [
  {
    text: "Save",
    onClick: () => {},
    type: "default",
  },
  {
    text: "Save and execute",
    onClick: () => {},
    divideTop: true,
    type: "default",
  },
  {
    text: "Save as draft",
    onClick: () => {},
    type: "default",
  },
  {
    text: "Delete",
    onClick: () => {},
    divideTop: true,
    type: "default",
  },
];

describe("ContextMenu component testing", () => {
  afterEach(cleanup);

  test("ContextMenu", async () => {
    const { asFragment, getByTestId } = renderWithThemeProvider(
      <ContextMenu contents={contents} />,
    );
    await act(async () => {
      fireEvent.click(getByTestId("icon-wrapper"));
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
