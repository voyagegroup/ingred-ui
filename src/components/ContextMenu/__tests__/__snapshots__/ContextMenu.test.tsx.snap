// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu component testing ContextMenu 1`] = `
<DocumentFragment>
  <button
    class="sc-bdvvtL hTa-DxM sc-eCImPb dPSedI"
    data-testid="icon-wrapper"
  >
    <div
      class="sc-gsDKAQ hIeZob"
    >
      <div
        class="sc-dkPtRN gozTXZ"
        size="18"
      >
        <svg
          viewBox="0 0 18 18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h18v18H0z"
            fill="none"
          />
          <path
            d="M11.5,3A1.5,1.5,0,1,0,13,4.5,1.5,1.5,0,0,0,11.5,3Zm0,10.5A1.5,1.5,0,1,0,13,15,1.5,1.5,0,0,0,11.5,13.5Zm0-5.25A1.5,1.5,0,1,0,13,9.75,1.5,1.5,0,0,0,11.5,8.25Z"
            fill="#0B82F4"
            transform="translate(-2.5 -0.75)"
          />
        </svg>
      </div>
    </div>
    <p
      class="sc-hKwDye gxqPEs"
      color="#0B82F4"
      font-size="14px"
    />
  </button>
  <div
    class="sc-kDTinF cvlHxF"
  >
    <div
      class="sc-crHmcD bXnuhj fade-transition-enter fade-transition-enter-active"
    >
      <div
        class="sc-iqseJM bDSzMX"
      />
    </div>
    <div
      class="sc-jrQzAO jahegW"
      data-popper-escaped="true"
      data-popper-placement="bottom-start"
      data-popper-reference-hidden="true"
      style="position: absolute; left: 0px; top: 0px;"
    >
      <div
        class="sc-jRQBWg hzzEuq"
      >
        <div
          class="sc-furwcr cQXUQE"
        >
          <p
            class="sc-hKwDye dATcrn sc-iCfMLu"
            color="#041C33"
            font-size="13px"
          >
            Save
          </p>
        </div>
        <hr
          class="sc-pVTFL fNLFDr"
          color="#E2E8EA"
          orientation="horizontal"
        />
        <div
          class="sc-furwcr cQXUQE"
        >
          <p
            class="sc-hKwDye dATcrn sc-iCfMLu"
            color="#041C33"
            font-size="13px"
          >
            Save and execute
          </p>
        </div>
        <div
          class="sc-furwcr cQXUQE"
        >
          <p
            class="sc-hKwDye dATcrn sc-iCfMLu"
            color="#041C33"
            font-size="13px"
          >
            Save as draft
          </p>
        </div>
        <hr
          class="sc-pVTFL fNLFDr"
          color="#E2E8EA"
          orientation="horizontal"
        />
        <div
          class="sc-furwcr cQXUQE"
        >
          <p
            class="sc-hKwDye dATcrn sc-iCfMLu"
            color="#041C33"
            font-size="13px"
          >
            Delete
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
